shared/src/sources/balancer/event_handler.rs

```diff
@@ -56,65 +62,119 @@ impl BalancerPoolRegistry {
     ) -> Result<Self> {
         let weighted_pool_factory = BalancerV2WeightedPoolFactory::deployed(&web3).await?;
         let two_token_pool_factory = BalancerV2WeightedPool2TokensFactory::deployed(&web3).await?;
+        let stable_pool_factory = BalancerV2StablePoolFactory::deployed(&web3).await?;
 
         let initial_pools = pool_initializer.initialize_pools().await?;
 
         let weighted_pool_updater = Mutex::new(EventHandler::new(
             web3.clone(),
             BalancerV2WeightedPoolFactoryContract(weighted_pool_factory),
-            PoolStorage::new(initial_pools.weighted_pools, pool_info.clone()),
+            PoolStorage::new(initial_pools.weighted_pools, vec![], pool_info.clone()),
             Some(initial_pools.fetched_block_number),
         ));
         let two_token_pool_updater = Mutex::new(EventHandler::new(
             web3.clone(),
             BalancerV2WeightedPool2TokensFactoryContract(two_token_pool_factory),
-            PoolStorage::new(initial_pools.weighted_2token_pools, pool_info),
+            PoolStorage::new(
+                initial_pools.weighted_2token_pools,
+                vec![],
+                pool_info.clone(),
+            ),
+            Some(initial_pools.fetched_block_number),
+        ));
+
+        let stable_pool_updater = Mutex::new(EventHandler::new(
+            web3.clone(),
+            BalancerV2StablePoolFactoryContract(stable_pool_factory),
+            PoolStorage::new(vec![], initial_pools.stable_pools, pool_info),
             Some(initial_pools.fetched_block_number),
         ));
 
         Ok(Self {
             weighted_pool_updater,
             two_token_pool_updater,
+            stable_pool_updater,
         })
     }
 
-    /// Retrieves `RegisteredWeightedPool`s from each Pool Store in the Registry and
+    /// Retrieves `RegisteredPool`s from each Pool Store in the Registry and
     /// returns the merged result.
     /// Primarily intended to be used by `BalancerPoolFetcher`.
     pub async fn get_pool_ids_containing_token_pairs(
```

Comment:
Will this be an issue now that a caller would request all the pool ids for the specified token pairs and wouldn't know which one of the `get_weighted_pools` or `get_stable_pools` methods to call it with returned pool IDs?

Reply:
Right, but I find the API a bit awkward:
```
let pool_ids = pools.get_pood_ids_containing_token_pairs(...);
let stable_pools = pools.get_stable_pools(pool_ids); // Note that there isn't 1 stable pool per ID
let weighted_pools = pools.get_weighted_pools(pool_ids); // Note that there isn't 1 weighted pool per ID.
```

It feels strange that we are joining stable pools and weighted pools in some places (for example `RegisteredPool`) but keeping them separate in other places (`get_stable_pools`/`get_weighted_pools`).

Reply:
Good point. I will see if I clean this up with a separation (with hopes of removing all the try_intos). 

I suppose we will need `get_stable_pool_ids_for_token_pairs` and `get_weighted_pool_ids_for_token_pairs` which isn't all that inconvenient anyway. The reason I opted for having one in the first place is that when we introduce yet another pool type (MetaStablePool) we will have three methods of each different kind in many places.

Reply:
I am not sure this is going to wind up being any nicer overall. We are going to have to split up the `PoolRegistry.pools_by_token` into two separate (but identical methods as well) and have a bunch of duplicated code. I will push a single commit with the changes so you can take a look.

Reply:
You could make `PoolRegistry` generic on the pool types. This would allow you to have a single method and only specify that bits that are different. This would also reduce code duplication and make adding new pools easier in the future.

Reply:
I have attempted to make some of this stuff generic, but whenever I put a `T` in place of `RegisteredXXXPool` it still expects me to specify something that `T` is known for...

Reply:
I guess then for all the "common" pool data, I would have to make a trait with methods accessing this and then state that `T: CommonPoolTrait`